File: HackerRank/30 Days of Code/Day13_AbstractClasses.cs
// Day 13: Abstract Classes
// https://www.hackerrank.com/challenges/30-abstract-classes/problem

using System;
using System.Collections.Generic;
using System.IO;
abstract class Book
{
    
    protected String title;
    protected  String author;
    
    public Book(String t,String a){
        title=t;
        author=a;
    }
    public abstract void display();


}

class MyBook : Book{
    int price;
    public MyBook(string t, string a, int p) : base(t, a){
        price = p;
    }
    public override void display(){
        Console.WriteLine("Title: {0}", title);
        Console.WriteLine("Author: {0}", author);
        Console.WriteLine("Price: {0}", price);
    }
}

class Solution {
    static void Main(String[] args) {
      String title=Console.ReadLine();
      String author=Console.ReadLine();
      int price=Int32.Parse(Console.ReadLine());
      Book new_novel=new MyBook(title,author,price);
      new_novel.display();
    }
}

// 추상 클래스를 상속받은 클래스 생성 및 사용
// C#에서 추상 메서드 생성 및 해당 클래스를 상속받아 실체화(override)

File: HackerRank/30 Days of Code/Day13_AbstractClasses.py
# Day 13: Abstract Classes
# https://www.hackerrank.com/challenges/30-abstract-classes/problem

from abc import ABCMeta, abstractmethod
class Book(object, metaclass=ABCMeta):
    def __init__(self,title,author):
        self.title=title
        self.author=author   
    @abstractmethod
    def display(): pass

class MyBook(Book):
    def __init__(self, t, a, p):
        Book.__init__(self, t, a)
        self.price = p
    def display(self):
        print("Title: " + self.title)
        print("Author: " + self.author)
        print("Price: " + str(self.price))

title=input()
author=input()
price=int(input())
new_novel=MyBook(title,author,price)
new_novel.display()

# 추상 클래스를 상속받은 클래스 생성 및 사용
# Python에서 추상 클래스를 사용하려면 from abc import *를 해줘야 하고, 추상 클래스에 metaclass로 ABCMeta를 지정해줘야 함에 유의